cns/ipampool/metrics/observer.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/Azure/azure-container-networking/crd/nodenetworkconfig/api/v1alpha"
+	"github.com/pkg/errors"
+)
+
+// Subnet ARM ID /subscriptions/$(SUB)/resourceGroups/$(GROUP)/providers/Microsoft.Network/virtualNetworks/$(VNET)/subnets/$(SUBNET)
+const subnetARMIDTemplate = "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s/subnets/%s"
+
+// ipPoolState is the current actual state of the CNS IP pool.
+type ipPoolState struct {
+	// allocatedToPods are the IPs CNS gives to Pods.
+	allocatedToPods int64
+	// available are the IPs in state "Available".
+	available int64
+	// currentAvailableIPs are the current available IPs: allocated - assigned - pendingRelease.
+	currentAvailableIPs int64
+	// expectedAvailableIPs are the "future" available IPs, if the requested IP count is honored: requested - assigned.
+	expectedAvailableIPs int64
+	// pendingProgramming are the IPs in state "PendingProgramming".
+	pendingProgramming int64
+	// pendingRelease are the IPs in state "PendingRelease".
+	pendingRelease int64
+	// requestedIPs are the IPs CNS has requested that it be allocated by DNC.
+	requestedIPs int64
+	// secondaryIPs are all the IPs given to CNS by DNC, not including the primary IP of the NC.
+	secondaryIPs int64
+}
+
+// metaState is the Monitor's configuration state for the IP pool.
+type metaState struct {
+	batch              int64
+	exhausted          bool
+	max                int64
+	primaryIPAddresses map[string]struct{}
+	subnet             string
+	subnetARMID        string
+	subnetCIDR         string
+}
+
+// NewLegacyMetricsObserver creates a closed functional scope which can be invoked to
+// observe the legacy IPAM pool metrics.
+//
+//nolint:lll // ignore line length
+func NewLegacyMetricsObserver(ctx context.Context, ipcli func() map[string]cns.IPConfigurationStatus, nnccli func(context.Context) (*v1alpha.NodeNetworkConfig, error), csscli func(context.Context) ([]v1alpha1.ClusterSubnetState, error)) func() error {
+	return func() error {
+		return observeMetrics(ctx, ipcli, nnccli, csscli)
+	}
+}
+
+// generateARMID uses the Subnet ARM ID format to populate the ARM ID with the metadata.
+// If either of the metadata attributes are empty, then the ARM ID will be an empty string.
+func generateARMID(nc *v1alpha.NetworkContainer) string {
+	subscription := nc.SubscriptionID
+	resourceGroup := nc.ResourceGroupID
+	vnetID := nc.VNETID
+	subnetID := nc.SubnetID
+
+	if subscription == "" || resourceGroup == "" || vnetID == "" || subnetID == "" {
+		return ""
+	}
+	return fmt.Sprintf(subnetARMIDTemplate, subscription, resourceGroup, vnetID, subnetID)
+}
+
+// observeMetrics observes the IP pool and updates the metrics. Blocking.
+//
+//nolint:lll // ignore line length
+func observeMetrics(ctx context.Context, ipcli func() map[string]cns.IPConfigurationStatus, nnccli func(context.Context) (*v1alpha.NodeNetworkConfig, error), csscli func(context.Context) ([]v1alpha1.ClusterSubnetState, error)) error {
```

Comment:
naming things `*cli` is kinda misleading because these are all just getter funcs no? Also, maybe you can define the getter functions type elsewhere to make this cleaner

Reply:
that's all true, but I intentionally went with anonymous types to make the footprint for this as small as possible (and not actually reusable, which is why we would use interfaces). Nothing can escape and it is very easy to rip out later.

Reply:
and `cli` is `client` which a getter func arguably is